### PR TITLE
Added methods to find broadcastAddress(es), useful in bindings for discovery.

### PR DESCRIPTION
**Pull request by mherwege**
Signed-off-by: Mark Herwege <mark.herwege@telenet.be>

### CODE REVIEW

**Comment by mherwege** [comment]
As suggested by @kaikreuzer in review of openHAB Niko Home Control binding [#1589](https://github.com/openhab/openhab2-addons/pull/1589).

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/net/NetUtil.java [hunk 1]
```diff
@@ -55,9 +60,53 @@ public static String getLocalIpv4HostAddress() {
...
     }
 
+    /**
+     * Get all broadcast addresses on the current host
+     *
+     * @return list of broadcast addresses
+     */
+    public static List<String> getAllBroadcastAddresses() {
+        try {
+            List<String> broadcastAddresses = new ArrayList<String>();
```
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/net/NetUtil.java]: Isn't a `LinkedList` better if we don't know the size (to reduce reallocations and copies)?

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/net/NetUtil.java [hunk 2]
```diff
@@ -55,9 +60,53 @@ public static String getLocalIpv4HostAddress() {
             }
             return hostAddress;
         } catch (SocketException ex) {
-            LOGGER.error("Could not retrieve network interface: " + ex.getMessage(), ex);
+            LOGGER.error("Could not retrieve network interface: {}", ex.getMessage(), ex);
```
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/net/NetUtil.java]: Good catch

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/net/NetUtil.java [hunk 3]
```diff
@@ -55,9 +60,53 @@ public static String getLocalIpv4HostAddress() {
...
             return null;
         }
     }
 
+    /**
+     * Get all broadcast addresses on the current host
+     *
+     * @return list of broadcast addresses
```
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/net/NetUtil.java]: If the `SocketException` is thrown you return `null`.
So, this should be stated in the JavaDoc.
But wouldn't it make sense to throw the exception, so the caller is notified about the error -- or if the caller cannot handle it at all (does he need to know about the error) return an empty list?

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/net/NetUtil.java [hunk 4]
```diff
@@ -55,9 +60,53 @@ public static String getLocalIpv4HostAddress() {
...
+     * Get the first candidate for a broadcast address
+     *
+     * @return broadcast address
+     */
+    public static String getBroadcastAddress() {
+        String broadcastAddress = null;
+        final List<String> broadcastAddresses = getAllBroadcastAddresses();
+        for (String current : broadcastAddresses) {
+            if (broadcastAddress != null) {
+                LOGGER.warn("Found multiple broadcast addresses - ignoring {}", broadcastAddress);
```
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/net/NetUtil.java]: The JavaDoc state "get the first candidate", so I don't see any reason to throw a warning if there are multiple candidates. The function should return the first one and not check if there is more then one, or should it?

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/net/NetUtil.java (hunk 3 above)
Comment (mherwege) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/net/NetUtil.java]: Returning an empty list, and documented.

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/net/NetUtil.java (hunk 1 above)
Comment (mherwege) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/net/NetUtil.java]: Replaced with a LinkedList.

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/net/NetUtil.java (hunk 4 above)
Comment (mherwege) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/net/NetUtil.java]: I used the same structure as in the `getLocalIpv4HostAddress()` method. It does not add value, so I removed it and just return the first one, or null if there is none. Null return is documented.

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/net/NetUtil.java [hunk 5]
```diff
@@ -55,7 +60,46 @@ public static String getLocalIpv4HostAddress() {
...
+        }
+    }
+
+    /**
+     * Get all broadcast addresses on the current host
+     *
+     * @return list of broadcast addresses, empty list if no broadcast addresses found
+     */
+    public static List<String> getAllBroadcastAddresses() {
+        List<String> broadcastAddresses = new LinkedList<String>();
```
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/net/NetUtil.java]: Hm, shouldn't the formatter replace `new LinkedList<String>()` be replaced by an `Diamond Operator`?

Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/net/NetUtil.java]: Hm, okay, seems it is not activated by the ESH formatter.

**Comment by maggu2810** [comment]
core and core tests succeeded, the error is not related

**Pull request merged**